### ISSUE TITLE
Consolidate Table of Content Components 

### ISSUE THREAD

**Issue by TylerAPfledderer** [ethereum/ethereum-org-website#8552]
A part of this [Discord discussion](https://discord.com/channels/714888181740339261/870313128242602004/1039930750805475438)

Investigate whether or not the ToC components need to be consolidated, at least between `UpgradeTableOfContents` and `StakingHomeTableOfContents` as they contain similar styling and display similar content.

**Comment by TylerAPfledderer** [comment]
@wackerow or @pettinarip please assign to me, as I have already begun exploring the issue.

**Comment by TylerAPfledderer** [comment]
@pettinarip a couple of initial observations:

There are three components for table of contents:
  - [TableOfContents](https://github.com/ethereum/ethereum-org-website/blob/dev/src/components/TableOfContents.tsx)
  - [UpgradeTableOfContents](https://github.com/ethereum/ethereum-org-website/blob/dev/src/components/UpgradeTableOfContents.tsx)
  - [StakingHomeTableOfContents](https://github.com/ethereum/ethereum-org-website/blob/dev/src/components/Staking/StakingHomeTableOfContents.tsx)

The second and third components are the most similar, with roughly the same styling and logic when generating the links. However, one current roadblock in combining them is the objects that provide the content for the links.
- Content sent to `StakingHomeTableOfContents` comes from an array within the page (i.e the [Learn](https://github.com/ethereum/ethereum-org-website/blob/2713f4cfdb4596f9d73ee0a9dc2a0661f5f58a0e/src/pages/learn/index.tsx#L196) and the [Staking](https://github.com/ethereum/ethereum-org-website/blob/2713f4cfdb4596f9d73ee0a9dc2a0661f5f58a0e/src/pages/staking/index.tsx#L307) page)
- Content sent to `UpgradeTableOfContents` comes from the static querying sent to the props of a template (i.e. the [Staking Template](https://github.com/ethereum/ethereum-org-website/blob/2713f4cfdb4596f9d73ee0a9dc2a0661f5f58a0e/src/templates/staking.tsx#L484)

The structure of these two arrays of objects are not the same and therefore the content is applied differently. It might be easier to convert the content structure used in `StakingHomeTableOfContents`.

With the `TableOfContents`, it has yet to be a migrated component to Chakra, but it's structure and styling is quite different.

Aside from how content is passed to these components, should `UpgradeTableOfContents` and `StakingHomeTableOfContents` be combined, would need to determine new names to distinguish the two remaining components.

**Comment by pettinarip** [comment]
Agreed with everything you described.

+1 on leaving `TableOfContents` as it is. It's pretty different from the other ones.
+1 on combining the other two, `UpgradeTableOfContents` & `StakingHomeTableOfContents`

In terms of data structure, I would lean toward using the [mdx `tableOfContents` data structure](https://www.gatsbyjs.com/plugins/gatsby-plugin-mdx/#graphql-mdx-node-structure) as you said.

One key difference I see between both is the way they generate the #`ids` for each item.
- `StakingHomeTableOfContents` expects you to pass the ids
- `UpgradeTableOfContents` generates them automatically from the `title` of each item

IMO we should auto-generate all of them but for the sake of simplicity of the scope of this issue, I would suggest giving support to both alternatives. Meaning to support an optional custom `id` in each item.
- if the item has an `id` => use that
- if not => generate it from the title text

cc @wackerow @minimalsm not sure if you have any opinion about this ^

In terms of naming, I would suggest keeping `UpgradeTableOfContents` for the consolidated new version. Have in mind that the TOC is something that probably is going to change/consolidate with the upcoming Design System.

**Comment by wackerow** [comment]
Yeah, this sounds good. It's that root staking page (not markdown) that made this more difficult. If we can optionally pass the custom id I think that would be a good solution

**Comment by TylerAPfledderer** [comment]
Thank you all for these thoughts. 🤘

I'll wait for PR #8703 to be reviewed and merged, then go ahead and tackle this further.

**Comment by corwintines** [comment]
@TylerAPfledderer looking at your last comment, is this something that can be tackled now?

**Comment by TylerAPfledderer** [comment]
@corwintines there is PR #8969 that has been waiting to be merged 😄
